examples/json.dx

```diff
@@ -0,0 +1,258 @@
+
+def join (extra: List a) (lists:n=>(List a)) : List a =
+    concat $ for i.
+        case ordinal i == (size n - 1) of
+             True -> lists.i
+             False -> lists.i <> extra
+
+
+data JValue = AsJValue String
+-- TODO - once Dex supports recursive ADT JValue becomes Value.
+
+data Value = 
+     AsObject (List (String & JValue))
+     AsArray (List JValue)
+     AsString String
+     AsFloat Float
+     AsInt Int
+
+interface ToJSON a
+  toJSON : a -> Value
+
+
+-- These three are private methods. Users should use type classes.
+def escape (x:JValue) : String =
+    (AsJValue y) = x
+    y
+
+
+def collapse (x:Value) : JValue =
+    AsJValue $ case x of
+         AsString y ->  "\"" <> y <> "\""
+         AsFloat y -> show y
+         AsInt y -> show y
+         AsObject (AsList _ y) ->
+              ("{" <> (join ", " $ for i.
+                 (k, v) = y.i 
+                 "\"" <> k <> "\"" <> ":" <> (escape v)) <> "}")
+         AsArray (AsList _ y) -> ("[" <> (join ", " $ for i. escape y.i) <> "]")
+
+def hide [ToJSON a] (x:a) : JValue =
+    collapse $ toJSON x
+
+instance Show Value 
+  show = \x. escape $ collapse x
+
+instance ToJSON String
+  toJSON = AsString
+  
+instance ToJSON Int
+  toJSON = AsInt
+  
+instance ToJSON Float
+  toJSON = AsFloat
+  
+instance ToJSON Value
+  toJSON = id
+  
+instance [ToJSON v] ToJSON ((Fin n) => v)
+  toJSON = \x . AsArray $ AsList _ $ for i. hide x.i
+
+instance [ToJSON v] ToJSON ((Fin n) => (String & v))
+  toJSON = \x . AsObject $ AsList _ $ for i. (fst x.i, hide $ snd x.i)
+
+instance [ToJSON v] ToJSON (List (String & v))
+  toJSON = \(AsList _ x) . toJSON x
+
+
+' ## Graph Grammars (Vega-Lite Spec)
+
+Options = List (String & String)
+
+data Mark =
+     AsMark String Options
+
+data EncodingType =
+     Quantitative
+     Nominal
+     Ordinal
+
+instance Show EncodingType
+         show = (\ x.
+              case x of
+                   Quantitative -> "quantitative"
+                   Nominal -> "nominal"
+                   Ordinal -> "ordinal")
+                   
+data Channel =
+     Y
+     X
+     Color
+     Tooltip
+     HREF
+     Row
+     Col
+     Size
+     
+instance Show Channel
+         show = (\ x.
+              case x of
+                   Y -> "y"
+                   X -> "x"
+                   Color -> "color"
+                   Tooltip -> "tooltip"
+                   HREF -> "href"
+                   Size -> "size"
+                   Row -> "row"
+                   Col -> "col")
+
+data Encoding key =
+     AsEncoding Channel key EncodingType Options
+
+def enc (c:Channel) (k:Int) (et: EncodingType) : Encoding key =
+    AsEncoding c (k@_) et mempty
+
+def mark (m:String) : Mark =
+    AsMark m mempty
+
+def optsList (x:Options) : List (String & Value) =
+    (AsList _  tab) = x
+    AsList _ $ for i. (fst tab.i, toJSON $ snd tab.i)
+
+
+data DataFrame key n value =
+     AsDataFrame (n => key => value) (key => String)
```

Comment:
Since `value` doesn't depend on `key` this is really a homogeneous data frame instead, right?

Reply:
Good point... it's heterogeneous because of the variant type for value. I guess ideally this would be a table where columns are of different types? I will have to think more if that is possible.

Reply:
Ohhh it's heterogeneous because you store all values as JSON! When we initially thought of data frames, we usually considered them as tables-of-records (`n=>{field1: Float & field2 : Int}`). A nice benefit of that is that those do have an efficient lowering, unlike this huge variant. But it would make it much harder to store the `column=>meta` association, because the `column` index set really indexes the fields of the record 🤔

Reply:
Yeah records are definitely the right solution to this. But I couldn't really figure out how to cleanly enumerate over fields. 

Although maybe I just need like a variant over Iso's instead of values. I guess the Meta could could do that work.

Reply:
Oh actually @danieldjohnson , maybe the cute solution is to have col metadata Iso Variant indexed and the columns in a Iso record. The metadata needs to provide a function to extract it's column (say into json)

Downside: The user would have to manually link them.  Same names?

Upside: homogenous data columns, variant named columns (no ordering), column enumeration.

Reply:
Tried a slightly different strategy. I think maybe flat dataframes are a hack and kind of unneeded in dex. To make graphing work, you can just construct the flat data representation along with the graph description. This can come from a record or from any other source. 

```haskell
n = Fin 100
df2 : {x1: n => Float &
        x2: n => Float &
        weight: n => Float &
        label : n => Class} 
...

chart2 = (AsVLDescriptor (pure Point) [("title", "Scatter")]
            [{title="X1", encodings=pureLs X, encType=Quantitative, rows=wrapCol #x1 df2},
             {title="X2", encodings=pureLs Y, encType=Quantitative, rows=wrapCol #x2 df2},
             {title="Weight", encodings=pureLs Size, encType=Quantitative, rows=wrapCol #weight df2},
             {title="Label", encodings=toList [pure Color, pure Tooltip], encType=Nominal, rows=wrapCol #label df2}])

:html showVega $ toJSON chart2

```

Reply:
Ooh, I like the most recent version!

I think that manipulating records will be much, much easier once we can enumerate over fields in userspace. So hopefully the "manually linking" downsides will no longer be a problem at that point.